File: data_structures/union_find/matlab/test_union_find.m
% test_union_find.m  (要素IDは0始まり)

uf = union_find('new', 5);
assert(union_find(uf, 'size') == 5);
assert(union_find(uf, 'components') == 5);
[~, b] = union_find(uf, 'connected', 0, 1);
assert(~b);

uf = union_find('new', 5);
[uf, ~] = union_find(uf, 'unite', 0, 1);
[uf, ~] = union_find(uf, 'unite', 1, 2);
[uf, b] = union_find(uf, 'connected', 0, 2);
assert(b);
[uf, b] = union_find(uf, 'connected', 0, 3);
assert(~b);
assert(union_find(uf, 'components') == 3);

uf = union_find('new', 5);
[uf, ~] = union_find(uf, 'unite', 0, 1);
[uf, ~] = union_find(uf, 'unite', 2, 3);
[uf, ~] = union_find(uf, 'unite', 0, 3);
[uf, b] = union_find(uf, 'connected', 1, 2);
assert(b);
assert(union_find(uf, 'components') == 2);
[uf, ok] = union_find(uf, 'unite', 0, 1);
assert(~ok);

disp('All tests passed.');

File: data_structures/union_find/python/test_union_find.py
import pytest
from union_find import UnionFind


def test_initial():
    uf = UnionFind(5)
    assert uf.size() == 5
    assert uf.components() == 5
    assert not uf.connected(0, 1)


def test_unite():
    uf = UnionFind(5)
    uf.unite(0, 1)
    uf.unite(1, 2)
    assert uf.connected(0, 2)
    assert not uf.connected(0, 3)
    assert uf.components() == 3


def test_transitivity():
    uf = UnionFind(5)
    uf.unite(0, 1)
    uf.unite(2, 3)
    uf.unite(0, 3)
    assert uf.connected(1, 2)
    assert uf.components() == 2
    assert not uf.unite(0, 1)


def test_bad_element_raises():
    uf = UnionFind(3)
    with pytest.raises(IndexError):
        uf.find(5)
